Public/Src/Tools/ServicePipDaemon/ServicePipDaemon.cs

```diff
@@ -0,0 +1,538 @@
+
+        /// <nodoc />
+        public ILogger Logger => m_logger;
+
+        /// <nodoc />
+        public ServicePipDaemon(IParser parser, DaemonConfig daemonConfig, ILogger logger, IIpcProvider rpcProvider = null, Client client = null)
+        {
+            Contract.Requires(daemonConfig != null);
+
+            Config = daemonConfig;
+            m_parser = parser;
+            ApiClient = client;
+            m_logger = logger;
+
+            rpcProvider = rpcProvider ?? IpcFactory.GetProvider();
+            m_server = rpcProvider.GetServer(Config.Moniker, Config);
+
+            m_etwLogger = new BuildXLBasedCloudBuildLogger(Config.Logger, Config.EnableCloudBuildIntegration);
+        }
+
+        /// <summary>
+        /// Starts to listen for client connections.  As soon as a connection is received,
+        /// it is placed in an action block from which it is picked up and handled asynchronously
+        /// (in the <see cref="ParseAndExecuteCommand"/> method).
+        /// </summary>
+        public void Start()
+        {
+            m_server.Start(this);
+        }
+
+        /// <summary>
+        /// Requests shut down, causing this daemon to immediately stop listening for TCP/IP
+        /// connections. Any pending requests, however, will be processed to completion.
+        /// </summary>
+        public void RequestStop()
+        {
+            m_server.RequestStop();
+        }
+
+        /// <summary>
+        /// Calls <see cref="RequestStop"/> then waits for <see cref="Completion"/>.
+        /// </summary>
+        public Task RequestStopAndWaitForCompletionAsync()
+        {
+            RequestStop();
+            return Completion;
+        }    
+
+        /// <inheritdoc />
+        public void Dispose()
+        {
+            m_server.Dispose();
+            ApiClient?.Dispose();
+            m_logger.Dispose();
+        }       
+
+        protected readonly CounterCollection<DaemonCounter> m_counters = new CounterCollection<DaemonCounter>();
+
+        protected enum DaemonCounter
+        {
+            /// <nodoc/>
+            [CounterType(CounterType.Stopwatch)]
+            ParseArgsDuration,
+
+            /// <nodoc/>
+            [CounterType(CounterType.Stopwatch)]
+            ServerActionDuration,
+
+            /// <nodoc/>
+            QueueDurationMs,
+        }
+
+        private async Task<IIpcResult> ParseAndExecuteCommand(IIpcOperation operation)
+        {
+            string cmdLine = operation.Payload;
+            m_logger.Verbose("Command received: {0}", cmdLine);
+            ConfiguredCommand conf;
+            using (m_counters.StartStopwatch(DaemonCounter.ParseArgsDuration))
+            {
+                conf = ParseArgs(cmdLine, m_parser);
+            }
+
+            IIpcResult result;
+            using (m_counters.StartStopwatch(DaemonCounter.ServerActionDuration))
+            {
+                 result = await conf.Command.ServerAction(conf, this);

```

Comment:
Just a comment for future use: it would be very helpful if we stick this `ServerActionDuration` time into `result`.